feat: raw_request and copy straight into available message

DonChangfoot opened the pull request:
`request`has been refactored to retrieve an available message from the
message bus and translate the incoming batch directly into it.

`raw_request` allows the user to handle the encoding themselves and give
the client a slice of bytes to send. This greatly improves performance
as only one copy into the available message is performed.

The README has also been updated with the correct data types.

Upstream has been updated.

Benchmark scripts updated to run `raw_request` as well.

DonChangfoot committed:
feat: implement raw_request and copy straight into available message

`request`has been refactored to retrieve an available message from the
message bus and translate the incoming batch directly into it.

`raw_request` allows the user to handle the encoding themselves and give
the client a slice of bytes to send. This greatly improves performance
as only one copy into the available message is performed.

The README has also been updated with the correct data types.

Upstream has been updated.

Benchmark scripts updated to run `raw_request` as well.

jorangreef commented on README.md:
````diff
@@ -40,7 +40,7 @@ One of the ways TigerBeetle achieves its performance is through batching. This i
 
 ```js
 const account = {
-    id: 137n, // u64
````
Good catch.

jorangreef commented on src/benchmark.ts:
```diff
@@ -4,20 +4,28 @@ import {
+  TransferFlags,
+  CreateTransfersError,
+  CommitTransfersError,
+  Operation
```
Almost wanted to say we should have a trailing comma here to benefit `zig fmt`, and then I realized... 😄

jorangreef commented on src/benchmark.ts:
```diff
@@ -48,55 +56,210 @@ const accountB: Account = {
+ * AMOUNT_OFFSET = 108 + 4 = 112
+ * TIMESTAMP_OFFSET = 112 + 8 = 120
+ */ 
+const encodeTransfer = (transfer: Transfer, offset: number, output: Buffer): void => {
```
I wish more Node.js APIs that work with buffers had function signatures like these. Most of them take a buffer, but few take an offset into the buffer, forcing slices and object allocations or interim allocations.

jorangreef commented on src/node.zig:
```diff
@@ -108,7 +113,13 @@ const Context = struct {
     message_bus: MessageBus,
     client: Client,
 
-    fn create(env: c.napi_env, allocator: *std.mem.Allocator, io: *IO, cluster: u128, configuration_raw: []const u8) !c.napi_value {
+    fn create(
+        env: c.napi_env,
+        allocator: *std.mem.Allocator,
+        io: *IO,
+        cluster: u128,
+        configuration_raw: []const u8,
+    ) !c.napi_value {
```
Great to fix these line lengths.

jorangreef commented on src/node.zig:
```diff
@@ -195,39 +209,44 @@ fn decode_from_object(comptime T: type, env: c.napi_env, object: c.napi_value) !
+        return translate.throw(env, "Batch is larger than the maximum message size.");
+    }
+
+    var results = std.mem.bytesAsSlice(T, output);
```
In Zig's std lib, `std.mem.bytesAsSlice` has this implementation:

```zig
/// Given a slice of bytes, returns a slice of the specified type
/// backed by those bytes, preserving pointer attributes.
pub fn bytesAsSlice(comptime T: type, bytes: anytype) BytesAsSliceReturnType(T, @TypeOf(bytes)) {
    // let's not give an undefined pointer to @ptrCast
    // it may be equal to zero and fail a null check
    if (bytes.len == 0) {
        return &[0]T{};
    }

    const cast_target = CopyPtrAttrs(@TypeOf(bytes), .Many, T);

    return @ptrCast(cast_target, bytes)[0..@divExact(bytes.len, @sizeOf(T))];
}
```

That final `@divExact()` requires that our `output` slice must be an exact multiple of `T`, so I think we need to slice on output before passing to `bytesAsSlice` to guarantee that, otherwise we may run into a safety-checked runtime panic for non-multiple sizes, even if the `body_length` does not overflow the buffer.

jorangreef commented on src/node.zig:
```diff
@@ -312,17 +438,25 @@ fn init(env: c.napi_env, info: c.napi_callback_info) callconv(.C) c.napi_value {
     return context;
 }
 
+/// This function decodes and validates an array of Node objects, one-by-one, directly into an 
+/// available message before requesting the client to send it.
```
Nice addition.

jorangreef commented on src/node.zig:
```diff
@@ -331,55 +465,147 @@ fn request(env: c.napi_env, info: c.napi_callback_info) callconv(.C) c.napi_valu
+        operation,
+        message.buffer[@sizeOf(Header)..],
+    ) catch |err| switch (err) {
         error.ExceptionThrown => {
             return null;
         },
```
We could make this a one-liner without using a block to enclose the `return null,`

jorangreef commented on src/node.zig:
```diff
@@ -331,55 +465,147 @@ fn request(env: c.napi_env, info: c.napi_callback_info) callconv(.C) c.napi_valu
+    const message = context.client.get_message() orelse {
+        translate.throw(
+            env,
+            "Message pool has been exhausted.",
```
```suggestion
            "Too many requests outstanding.",
```

jorangreef commented on src/node.zig:
```diff
@@ -331,55 +465,147 @@ fn request(env: c.napi_env, info: c.napi_callback_info) callconv(.C) c.napi_valu
+/// copy directly into an available message. No validation of the encoded data is performed except
+/// that it will fit into the message buffer.
+fn raw_request(env: c.napi_env, info: c.napi_callback_info) callconv(.C) c.napi_value {
+    var argc: usize = 20;
+    var argv: [20]c.napi_value = undefined;
```
I think there are a few places within `node.zig` where we allocate space on the stack like this for `20` argv values. However, we can be more efficient and allocate only as many as we decide we want to get back from `napi_get_cb_info()`. We just set `argc` to the number of args we want, and `argv[N]` to the same, and `napi_get_cb_info()` will only copy that many (if there are more it will stop at how many we expect).

jorangreef replied:
For example, in this case, we would want to do:

```zig
var argc: usize = 4;
var argv: [4]c.napi_value = undefined;
```

`napi_get_cb_info()` will then set `argc` to the actual number of args passed by the user, e.g. if more than 4 so we can still do our error handling below for `Function request() requires 4 arguments exactly.` as we do.

jorangreef commented on src/node.zig:
```diff
@@ -331,55 +465,147 @@ fn request(env: c.napi_env, info: c.napi_callback_info) callconv(.C) c.napi_valu
+    const message = context.client.get_message() orelse {
+        translate.throw(
+            env,
+            "Message pool has been exhausted.",
```
```suggestion
            "Too many requests outstanding.",
```

jorangreef replied:
I think saying something like this helps the user decide how they can fix this from happening on their end, without needing to know about our message pool internals.

jorangreef commented on src/node.zig:
```diff
@@ -331,55 +465,147 @@ fn request(env: c.napi_env, info: c.napi_callback_info) callconv(.C) c.napi_valu
+        message.buffer[@sizeOf(Header)..],
+        "raw_batch",
+    ) catch |err| switch (err) {
+        error.ExceptionThrown => {
+            return null;
+        },
```
```suggestion
        error.ExceptionThrown => return null,
```

jorangreef commented on src/node.zig:
```diff
@@ -331,55 +465,147 @@ fn request(env: c.napi_env, info: c.napi_callback_info) callconv(.C) c.napi_valu
+    return null;
+}
+
+fn create_client_error(env: c.napi_env, client_error: ClientError) !c.napi_value {
+    return switch (client_error) {
+        error.TooManyOutstandingRequests => try translate.create_error(
+            env,
+            "Too many outstanding requests - message pool exhausted.",
+        ),
+    };
+}
```
Nice!

jorangreef commented on src/translate.zig:
```diff
@@ -18,14 +18,15 @@ pub fn register_function(
 }
 
-pub fn throw(env: c.napi_env, comptime message: [:0]const u8) error{ExceptionThrown} {
+const TranslationErrors = error{ExceptionThrown};
```
I think in the Zig std lib etc. they just use the singular for the name of the error set type, e.g. here we could do `TranslationError`.

jorangreef replied:
Thanks for making these error sets explicit, rather than inferred.

jorangreef commented on src/translate.zig:
```diff
@@ -153,6 +159,26 @@ pub fn bytes_from_object(env: c.napi_env, object: c.napi_value, comptime length:
+        return throw(env, key ++ " exceeds max message size.");
+    }
+
+    // copy this out of V8 as the underlying data lifetime is not guaranteed.
```
```suggestion
    // Copy this out of V8 as the underlying data lifetime is not guaranteed:
```

jorangreef commented on src/translate.zig:
```diff
@@ -153,6 +159,26 @@ pub fn bytes_from_object(env: c.napi_env, object: c.napi_value, comptime length:
+    comptime key: [:0]const u8,
+) !usize {
+    const data = try slice_from_value(env, buffer, key);
+    if (data.len < 1) {
+        return throw(env, key ++ " must not be empty.");
+    }
```
We could add a similar check to `client.zig` to require that a request body must be provided (at least for now), e.g. `body_size > 0`. Later we can allow some operations to not specify a body.

DonChangfoot replied:
@jorangreef  I'll make a change in the `tigerbeetle` repo and update the upstream in another PR.

jorangreef commented on src/node.zig:
```diff
@@ -331,55 +465,147 @@ fn request(env: c.napi_env, info: c.napi_callback_info) callconv(.C) c.napi_valu
+        argv[1] = globals.napi_undefined;
+    }
+
+    translate.call_function(env, scope, napi_callback, argc, argv[0..]) catch {
+        translate.throw(env, "Failed to call JS results callback.") catch return;
+    };
```
It's possible for the user's own JS callback function to throw an exception. If that happens, we'll land up here since the result checked by https://github.com/coilhq/tigerbeetle-node/blob/6b990b66f8696a72acb6ab96d5f7016611863fe7/src/translate.zig#L443 will then not be `.napi_ok`.

We should rather tweak L443 in `translate.zig` to expect and handle this as ok, so that we don't throw an exception here. If it's the user throwing an exception, then that's fine from our point of view, it's all within their application. For example, we may have called their callback, and that called something synchronously and so on until eventually that ended up with a user exception firing.

jorangreef reviewed (changes requested):
Awesome work @DonChangfoot!

DonChangfoot committed:
fix: review changes

fix a memory leak where we would not free up callback
references we created in V8 if an error occurred processing a request.

size `argv` to exactly how many arguments we expect.

we now don't throw a JS error if the user callback has already thrown
its own error.

jorangreef reviewed (approved):
LATM!
